src/next_array.rs

```diff
@@ -0,0 +1,254 @@
+use core::mem::{self, MaybeUninit};
+
+/// An array of at most `N` elements.
+struct ArrayBuilder<T, const N: usize> {
+    /// The (possibly uninitialized) elements of the `ArrayBuilder`.
+    ///
+    /// # Safety
+    ///
+    /// The elements of `arr[..len]` are valid `T`s.
+    arr: [MaybeUninit<T>; N],
+
+    /// The number of leading elements of `arr` that are valid `T`s, len <= N.
+    len: usize,
+}
+
+impl<T, const N: usize> ArrayBuilder<T, N> {
+    /// Initializes a new, empty `ArrayBuilder`.
+    pub fn new() -> Self {
+        // SAFETY: The safety invariant of `arr` trivially holds for `len = 0`.
+        Self {
+            arr: [(); N].map(|_| MaybeUninit::uninit()),
+            len: 0,
+        }
+    }
+
+    /// Pushes `value` onto the end of the array.
+    ///
+    /// # Panics
+    ///
+    /// This panics if `self.len >= N`.
+    #[inline(always)]
+    pub fn push(&mut self, value: T) {
+        // PANICS: This will panic if `self.len >= N`.
+        let place = &mut self.arr[self.len];
+        // SAFETY: The safety invariant of `self.arr` applies to elements at
+        // indices `0..self.len` — not to the element at `self.len`. Writing to
+        // the element at index `self.len` therefore does not violate the safety
+        // invariant of `self.arr`. Even if this line panics, we have not
+        // created any intermediate invalid state.
+        *place = MaybeUninit::new(value);
+        // PANICS: This cannot panic, since `self.len < N <= usize::MAX`.
+        // `0..self.len` are valid. Due to the above write, the element at
+        // `self.len` is now also valid. Consequently, all elements at indicies
+        // `0..(self.len + 1)` are valid, and `self.len` can be safely
+        // incremented without violating `self.arr`'s invariant. It is fine if
+        // this increment panics, as we have not created any intermediate
+        // invalid state.
```

Comment:
A few concerns with this:
- Parts of this comment are actually a safety comment, not a panics comment
- The safety comment aspects of this are insufficient

Here's a suggested update:

```suggestion
        // Lemma: `self.len < N`. By invariant, `self.len <= N`. Above, we index
        // into `self.arr`, which has size `N`, at index `self.len`. If `self.len == N`
        // at that point, that index would be out-of-bounds, and the index
        // operation would panic. Thus, `self.len != N`, and since `self.len <= N`,
        // that means that `self.len < N`.
        //
        // PANICS: Since `self.len < N`, and since `N <= usize::MAX`,
        // `self.len + 1 <= usize::MAX`, and so `self.len += 1` will not
        // overflow. Overflow is the only panic condition of `+=`.
        //
        // SAFETY:
        // - We are required to uphold the invariant that `self.len <= N`.
        //   Since, by the preceding lemma, `self.len < N` at this point in the
        //   code, `self.len += 1` results in `self.len <= N`.
        // - We are required to uphold the invariant that `self.arr[..self.len]`
        //   are valid instances of `T`. Since this invariant already held when
        //   this method was called, and since we only increment `self.len`
        //   by 1 here, we only need to prove that the element at
        //   `self.arr[self.len]` (using the value of `self.len` before incrementing)
        //   is valid. Above, we construct `place` to point to `self.arr[self.len]`,
        //   and then initialize `*place` to `MaybeUninit::new(value)`, which is
        //   a valid `T` by construction.
```